src/Arrays/MPIStateArrays.jl

```diff
@@ -20,6 +20,13 @@ Base.similar(::Type{A}, ::Type{FT}, dims...) where {A <: Array, FT} =
     Base.similar(::Type{A}, ::Type{FT}, dims...) where {A <: CuArray, FT} =
         similar(CuArray{FT}, dims...)
 end
+@init @require CUDAnative = "be33ccc6-a3ff-5ff2-a52e-74243cff1e17" begin
+    using .CUDAnative
+    using Adapt
+    # allows passing MPIStateArrays to GPU kernels
+    Adapt.adapt_storage(to::CUDAnative.Adaptor, a::MPIStateArray) =
+        adapt(to, a.data)
```

Comment:
So the goal is to unpack the MPIStateArray and not to provide a variant that works on the GPU?

I would have expected something like
```
Adapt.adapt_structure(to, x::MPIStateArray) = MPIStateArray(adapt(to, a.data))
```

Which we should probably still define to make it easier to copy a MPIStateArray to and from the GPU with `adapt(CuArray, MPIStateArray)`.

The way you are currently doing it will lead to CPU and GPU seeing different datastructures.

Reply:
The goal is to be able to pass an `MPIStateArray` to a kernel, unpacking it was just the easiest way to achieve this. Of course, having a variant that works on the GPU would be ideal. However, since `MPIStateArray` contains not only data but a bunch of other stuff, most of which looks like it's not GPU-compatible (MPI communicator, `Arrays` of MPI requests, buffers, etc.) I don't think it is as easy as doing `MPIStateArray(adapt(to, a.data))`. In practice, we only use the indexing functionality inside kernels. In the long run it seems like #690 may address this.

Reply:
I guess having an implicit conversion from the `MPIStateArray` to the underlying data array would also be a solution to the annoyances that inspired this PR. What this PR does is half of the implicit conversion, really.

Reply:
Yeah we could define an KernelAbstraction `Adaptor` and use that to do the unpacking/transforming it into a `StateArray`